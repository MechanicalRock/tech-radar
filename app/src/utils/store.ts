import * as reducers from '../reducers';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';

const store = createStore(
  combineReducers<ReduxState>(reducers),
  applyMiddleware(thunk)
);

export default store;
