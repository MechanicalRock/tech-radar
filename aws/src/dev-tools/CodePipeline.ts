import { Pipeline, GitHubSourceAction } from '@aws-cdk/aws-codepipeline';
import {
  PipelineBuildAction,
  PipelineProject,
  LinuxBuildImage,
  PipelineBuildActionProps
} from '@aws-cdk/aws-codebuild';
import { Stack, SecretParameter } from '@aws-cdk/cdk';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';

export interface CodePipelineProps {
  stack: Stack;
  logicalName: string;
  pipelineName: string;
}

export class CodePipeline {
  private pipeline: Pipeline;
  private codeBuildRole: Role;

  constructor(private props: CodePipelineProps) {
    this.create();
  }

  create() {
    // Create the pipeline
    this.pipeline = new Pipeline(this.props.stack, this.props.logicalName, {
      pipelineName: this.props.pipelineName,
      restartExecutionOnUpdate: true
    });

    this.createCodeBuildRole();
    this.createSourceStage();
    this.createInfraStage();
    this.createDevStage();
    this.createProdStage();
  }

  private createCodeBuildRole(): void {
    this.codeBuildRole = new Role(this.props.stack, 'CodeBuildRole', {
      assumedBy: new ServicePrincipal('codebuild.amazonaws.com')
    });
    this.codeBuildRole.attachManagedPolicy(
      'arn:aws:iam::aws:policy/AdministratorAccess'
    );
  }

  private createProdStage(): void {
    // Prod stage and build & test action
    const stage = this.pipeline.addStage('ProdStage');
    const project = this.createCodeBuildProject('DeployToProd');

    this.createCodeBuildAction('prodDeployAction', {
      project,
      stage
    });
  }

  private createInfraStage(): void {
    // Deploy Infra
    const stage = this.pipeline.addStage('InfraStage');
    const project = this.createCodeBuildProject(
      'DeployInfra',
      'infra_buildspec.yml'
    );

    this.createCodeBuildAction('DeployInfraAction', {
      project,
      stage
    });
  }

  private createDevStage(): void {
    // Dev stage and build & test action
    const stage = this.pipeline.addStage('DevStage');
    const project = this.createCodeBuildProject(
      'BuildAndTest',
      'app_buildspec.yml'
    );

    this.createCodeBuildAction('BuildAndTestAction', {
      project,
      stage
    });
  }

  private createCodeBuildProject(
    logicalName: string,
    buildSpec?: string
  ): PipelineProject {
    return new PipelineProject(this.props.stack, logicalName, {
      environment: {
        buildImage: LinuxBuildImage.UBUNTU_14_04_NODEJS_10_1_0
      },
      role: this.codeBuildRole,
      buildSpec
    });
  }

  private createCodeBuildAction(
    actionName: string,
    props: PipelineBuildActionProps
  ): void {
    new PipelineBuildAction(this.props.stack, actionName, props);
  }

  private createSourceStage(): void {
    const stage = this.pipeline.addStage('SourceStage');

    const { value: oauthToken } = new SecretParameter(
      this.props.stack,
      'GitHubOAuthToken',
      { ssmParameter: 'github-oauth-token' }
    );

    new GitHubSourceAction(this.props.stack, 'GitHub_Source', {
      stage,
      owner: 'MechanicalRock',
      repo: 'tech-radar',
      oauthToken
    });
  }
}
