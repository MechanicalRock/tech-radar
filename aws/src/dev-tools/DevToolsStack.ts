import { Stack, App } from '@aws-cdk/cdk';
import { CodePipeline } from './CodePipeline';

export class DevToolsStack extends Stack {
  constructor(app: App, stackName: string) {
    super(app, stackName);
    this.create();
  }

  create() {
    // Create a code pipeline
    new CodePipeline({
      stack: this,
      logicalName: 'CodePipeline',
      pipelineName: 'code-pipeline'
    });
  }
}
