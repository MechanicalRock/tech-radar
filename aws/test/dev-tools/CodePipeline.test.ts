import { Stack } from '@aws-cdk/cdk';
import * as yaml from '../utils/Yaml';
import { CodePipeline } from '../../src/dev-tools/CodePipeline';

describe('CodePipeline', () => {
    it('should match the template snapshot', () => {
        // Given
        const stack = new Stack();

        // When
        new CodePipeline({
            stack,
            logicalName: 'CodePipeline',
            pipelineName: 'code-pipeline',
        });

        // Then
        expect(yaml.toYamlString(stack.toCloudFormation())).toMatchSnapshot();
    })
})