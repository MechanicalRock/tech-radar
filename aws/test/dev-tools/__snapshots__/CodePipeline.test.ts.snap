// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodePipeline should match the template snapshot 1`] = `
"Resources:
  CodePipelineArtifactsBucketF1E925CF:
    Type: 'AWS::S3::Bucket'
    DeletionPolicy: Retain
  CodePipelineRoleB3A660B4:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: 'sts:AssumeRole'
            Effect: Allow
            Principal:
              Service: codepipeline.amazonaws.com
        Version: '2012-10-17'
  CodePipelineRoleDefaultPolicy8D520A8D:
    Type: 'AWS::IAM::Policy'
    Properties:
      PolicyDocument:
        Statement:
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
          - Action:
              - 'codebuild:BatchGetBuilds'
              - 'codebuild:StartBuild'
              - 'codebuild:StopBuild'
            Effect: Allow
            Resource:
              'Fn::GetAtt':
                - BuildAndTestB9A2F419
                - Arn
          - Action:
              - 'codebuild:BatchGetBuilds'
              - 'codebuild:StartBuild'
              - 'codebuild:StopBuild'
            Effect: Allow
            Resource:
              'Fn::GetAtt':
                - DeployToProd5A783BE8
                - Arn
        Version: '2012-10-17'
      PolicyName: CodePipelineRoleDefaultPolicy8D520A8D
      Roles:
        - Ref: CodePipelineRoleB3A660B4
  CodePipelineB74E5936:
    Type: 'AWS::CodePipeline::Pipeline'
    Properties:
      RoleArn:
        'Fn::GetAtt':
          - CodePipelineRoleB3A660B4
          - Arn
      Stages:
        - Actions:
            - ActionTypeId:
                Category: Source
                Owner: ThirdParty
                Provider: GitHub
                Version: '1'
              Configuration:
                Owner: MechanicalRock
                Repo: tech-radar
                Branch: master
                PollForSourceChanges: false
              InputArtifacts: []
              Name: GitHub_Source
              OutputArtifacts:
                - Name: Artifact_GitHub_Source
              RunOrder: 1
          Name: SourceStage
        - Actions:
            - ActionTypeId:
                Category: Build
                Owner: AWS
                Provider: CodeBuild
                Version: '1'
              Configuration:
                ProjectName:
                  Ref: BuildAndTestB9A2F419
              InputArtifacts:
                - Name: Artifact_GitHub_Source
              Name: BuildAndTestAction
              OutputArtifacts:
                - Name: Artifact_BuildAndTestAction
              RunOrder: 1
          Name: DevStage
        - Actions:
            - ActionTypeId:
                Category: Build
                Owner: AWS
                Provider: CodeBuild
                Version: '1'
              Configuration:
                ProjectName:
                  Ref: DeployToProd5A783BE8
              InputArtifacts:
                - Name: Artifact_BuildAndTestAction
              Name: prodDeployAction
              OutputArtifacts:
                - Name: Artifact_prodDeployAction
              RunOrder: 1
          Name: ProdStage
      ArtifactStore:
        Location:
          Ref: CodePipelineArtifactsBucketF1E925CF
        Type: S3
      Name: code-pipeline
      RestartExecutionOnUpdate: true
    DependsOn:
      - CodePipelineRoleB3A660B4
      - CodePipelineRoleDefaultPolicy8D520A8D
  CodeBuildRole728CBADE:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: 'sts:AssumeRole'
            Effect: Allow
            Principal:
              Service: codebuild.amazonaws.com
        Version: '2012-10-17'
      ManagedPolicyArns:
        - 'arn:aws:iam::aws:policy/AdministratorAccess'
  CodeBuildRoleDefaultPolicy829527DE:
    Type: 'AWS::IAM::Policy'
    Properties:
      PolicyDocument:
        Statement:
          - Action:
              - 'logs:CreateLogGroup'
              - 'logs:CreateLogStream'
              - 'logs:PutLogEvents'
            Effect: Allow
            Resource:
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: BuildAndTestB9A2F419
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: BuildAndTestB9A2F419
                    - ':*'
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
          - Action:
              - 'logs:CreateLogGroup'
              - 'logs:CreateLogStream'
              - 'logs:PutLogEvents'
            Effect: Allow
            Resource:
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: DeployToProd5A783BE8
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: DeployToProd5A783BE8
                    - ':*'
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
        Version: '2012-10-17'
      PolicyName: CodeBuildRoleDefaultPolicy829527DE
      Roles:
        - Ref: CodeBuildRole728CBADE
  GitHubSourceWebhookResourceB32EADBE:
    Type: 'AWS::CodePipeline::Webhook'
    Properties:
      Authentication: GITHUB_HMAC
      AuthenticationConfiguration: {}
      Filters:
        - JsonPath: $.ref
          MatchEquals: 'refs/heads/{Branch}'
      TargetAction: GitHub_Source
      TargetPipeline:
        Ref: CodePipelineB74E5936
      TargetPipelineVersion: 1
      RegisterWithThirdParty: true
  BuildAndTestB9A2F419:
    Type: 'AWS::CodeBuild::Project'
    Properties:
      Artifacts:
        Type: CODEPIPELINE
      Environment:
        ComputeType: BUILD_GENERAL1_SMALL
        Image: 'aws/codebuild/nodejs:10.1.0'
        PrivilegedMode: false
        Type: LINUX_CONTAINER
      ServiceRole:
        'Fn::GetAtt':
          - CodeBuildRole728CBADE
          - Arn
      Source:
        Type: CODEPIPELINE
  DeployToProd5A783BE8:
    Type: 'AWS::CodeBuild::Project'
    Properties:
      Artifacts:
        Type: CODEPIPELINE
      Environment:
        ComputeType: BUILD_GENERAL1_SMALL
        Image: 'aws/codebuild/nodejs:10.1.0'
        PrivilegedMode: false
        Type: LINUX_CONTAINER
      ServiceRole:
        'Fn::GetAtt':
          - CodeBuildRole728CBADE
          - Arn
      Source:
        Type: CODEPIPELINE
"
`;
